### DISCUSSION TITLE
Volumetric mesh cable-driven arm

### DISCUSSION THREAD

**Discussion by camilla-agabiti** [SofaDefrost/SoftRobots#137] [Debate - Technical topics]
Hi everyone,

I am building up a simulation of a cable-driven soft robot that is supposed to perform whole-arm grasping. 
In a first design, the arm has a conical shape and it is actuated by 3 cables; the cables are arranged inside the silicon body of the arm at an angular distance of 120° from each other. I am in the phase of making the CAD of the object (conical arm) and consequently generate the volumetric 3-D mesh to import in the simulation.
I have a doubt related to the cables arrangement; in particular, since the cables are constrained inside the robot and embedded in the silicon body, must the CAD of the object include the three cylindrical cavities allowing the cables pass inside the silicon arm? 

I would be very grateful to have an answer from you.

Thank you!

Camilla

**Comment by EulalieCoevoet** [comment]
Hi Camilla, 

Thanks for the question. The answer is no. We use a concept of mappings to allow the passing points of the cable to be located in the middle of an FEM element (you can see section 4.1 of [this paper](https://hal.inria.fr/hal-01649355/document) for more details).
You just need to define the position of these points and use the component `BarycentricMapping`. For instance in [this example](https://github.com/SofaDefrost/SoftRobots/blob/master/docs/sofapython3/examples/component/constraint/CableConstraint/Finger.py3):

```python
 cable = finger.addChild('cable')

 # This creates a MechanicalObject, a component holding the degree of freedom of our
 # mechanical modelling. In the case of a cable it is a set of positions specifying
 # the points where the cable is passing by.
 cable.addObject('MechanicalObject',
                  position=[
                      [-17.5, 12.5, 2.5],
                      [-32.5, 12.5, 2.5],
                      [-47.5, 12.5, 2.5],
                      [-62.5, 12.5, 2.5],
                      [-77.5, 12.5, 2.5],

                      [-85.5, 12.5, 6.5],
                      [-85.5, 12.5, 8.5],
                      [-83.5, 12.5, 4.5],
                      [-83.5, 12.5, 10.5],

                      [-77.5, 12.5, 12.5],
                      [-62.5, 12.5, 12.5],
                      [-47.5, 12.5, 12.5],
                      [-32.5, 12.5, 12.5],
                      [-17.5, 12.5, 12.5]])

 # Create a CableConstraint object. The indices are referring to the MechanicalObject's positions.
 cable.addObject('CableConstraint',
                  indices=list(range(0, 14)),
                  minForce=0,  # Set that the cable can't push
                  pullPoint=[0.0, 12.5, 2.5])

 # This creates a BarycentricMapping. A BarycentricMapping is a key element as it will create a bidirectional link
 # between the cable's DoFs and the finger's ones, so that movements of the cable's DoFs will be mapped
 # to the finger and vice-versa;
 cable.addObject('BarycentricMapping')
```

**Reply by camilla-agabiti** [reply]
Hi Eulalie,

Thank you very much for your reply! I am starting to actively use this modeling tool and also the existing design of your trunk-line robot to make my own simulations. It would be great if then we could have some exchanges about our works.

Thank you again
Camilla
